src/treesitter.rs

```diff
@@ -21,3 +31,359 @@ impl From<&'_ Node<'_>> for Match {
+                    continue;
+                }
+                let single_captured_node = this_capture.node;
+                if all_fields
+                    .filter
+                    .as_ref()
+                    .map_or(true, |filter| filter.call(&single_captured_node))
+                {
+                    *all_fields.next_capture = Some(CaptureInfo {
+                        node: single_captured_node,
+                        pattern_index: match_.pattern_index,
+                    });
+                    return;
+                }
+            }
+            *all_fields.next_capture = None;
+        });
+    }
+
+    fn get<'this>(&'this self) -> Option<&'this Self::Item> {
+        let next_capture = self.borrow_next_capture();
+        // SAFETY: I think this is ok as long as CaptureInfo isn't
+        // Copy/Clone?
+        // Since at that point there's no way for the "inner"
+        // CaptureInfo's contents to "outlive" the returned reference?
+        // Did this because otherwise was running into not being able
+        // to express that the "real" Item type for this trait (I think)
+        // should be CaptureInfo<'this>, not CaptureInfo<'a>
+        let next_capture: &'this Option<CaptureInfo<'tree>> =
+            unsafe { mem::transmute(next_capture) };
+        next_capture.as_ref()
+    }
+}
+
+#[self_referencing]
+pub struct Matches<'a, 'text: 'a, 'tree: 'a> {
+    text: RopeOrSlice<'text>,
+    query_cursor: QueryCursor,
+    query: &'a Query,
+    tree: Cow<'tree, Tree>,
+    #[borrows(text, mut query_cursor, query, tree)]
+    #[covariant]
+    matches_iterator: QueryMatches<'this, 'this, 'this, RopeOrSlice<'this>>,
+    #[borrows(tree)]
+    #[covariant]
+    next_match: Option<QueryMatch<'this, 'this>>,
+}
+
+pub fn get_matches<'a, 'text, 'tree>(
+    language: Language,
+    text: impl Into<RopeOrSlice<'text>>,
+    query: &'a Query,
+    tree: Option<&'tree Tree>,
+) -> Matches<'a, 'text, 'tree> {
+    let text = text.into();
+    let query_cursor = QueryCursor::new();
+    let tree: Cow<'tree, Tree> = tree.map_or_else(
+        || Cow::Owned(text.parse(&mut get_parser(language), None).unwrap()),
+        Cow::Borrowed,
+    );
+    Matches::new(
+        text,
+        query_cursor,
+        query,
+        tree,
+        |text, query_cursor, query, tree| query_cursor.matches(query, tree.root_node(), *text),
+        |_| None,
+    )
+}
+
+impl<'a, 'text, 'tree> StreamingIterator for Matches<'a, 'text, 'tree> {
+    type Item = QueryMatch<'a, 'tree>;
+
+    fn advance(&mut self) {
+        self.with_mut(|all_fields| {
+            *all_fields.next_match = all_fields.matches_iterator.next();
+        });
+    }
+
+    fn get<'this>(&'this self) -> Option<&'this Self::Item> {
+        let next_match = self.borrow_next_match();
+        // SAFETY: Not as sure on this one?
+        let next_match: &'this Option<QueryMatch<'a, 'tree>> =
+            unsafe { mem::transmute(next_match) };
```

Comment:
Same "technique" here (as with `Captures` above) for avoiding lifetime errors

No idea but I'd be surprised if this isn't "sound"